cmd/lookup/main.go

```diff
@@ -35,53 +46,66 @@ func main() {
 			continue
 		}
 
-		ref, err := registry.ParseRef(img)
-		if err != nil {
-			panic(err)
-		}
-
-		var obj any
-		if opts == zeroOpts {
-			// if we have no explicit type and didn't request a HEAD, invoke SynthesizeIndex instead of Lookup
-			obj, err = registry.SynthesizeIndex(ctx, ref)
-			if err != nil {
-				panic(err)
-			}
-		} else {
-			r, err := registry.Lookup(ctx, ref, &opts)
+		do := func(opts registry.LookupOptions) {
+			ref, err := registry.ParseRef(img)
 			if err != nil {
 				panic(err)
 			}
-			if r != nil {
-				desc := r.Descriptor()
-				if opts.Head {
-					obj = desc
-				} else {
-					b, err := io.ReadAll(r)
-					if err != nil {
-						r.Close()
-						panic(err)
-					}
-					if opts.Type == registry.LookupTypeManifest {
-						// if it was a manifest lookup, cast the byte slice to json.RawMessage so we get the actual JSON (not base64)
-						obj = json.RawMessage(b)
-					} else {
-						obj = b
-					}
-				}
-				err = r.Close()
+
+			var obj any
+			if opts == zeroOpts {
+				// if we have no explicit type and didn't request a HEAD, invoke SynthesizeIndex instead of Lookup
+				obj, err = registry.SynthesizeIndex(ctx, ref)
 				if err != nil {
 					panic(err)
 				}
 			} else {
-				obj = nil
+				r, err := registry.Lookup(ctx, ref, &opts)
+				if err != nil {
+					panic(err)
+				}
+				if r != nil {
+					desc := r.Descriptor()
+					if opts.Head {
+						obj = desc
+					} else {
+						b, err := io.ReadAll(r)
+						if err != nil {
+							r.Close()
+							panic(err)
+						}
+						if opts.Type == registry.LookupTypeManifest {
+							// if it was a manifest lookup, cast the byte slice to json.RawMessage so we get the actual JSON (not base64)
+							obj = json.RawMessage(b)
+						} else {
+							obj = b
+						}
+					}
+					err = r.Close()
+					if err != nil {
+						panic(err)
+					}
+				} else {
+					obj = nil
+				}
+			}
+
+			e := json.NewEncoder(os.Stdout)
+			e.SetIndent("", "\t")
+			if err := e.Encode(obj); err != nil {
+				panic(err)
 			}
 		}
 
-		e := json.NewEncoder(os.Stdout)
-		e.SetIndent("", "\t")
-		if err := e.Encode(obj); err != nil {
-			panic(err)
+		if parallel {
+			wg.Add(1)
```

Comment:
Should there be a limit on how many should be done in parallel?

Reply:
If this were meant for more than testing extreme concurrency, yes, probably, but the purpose of the new flag on this internal tool is specifically to test massive parallelism (as in my description, I had to get to 10k concurrent requests before I could actually observe a performance difference).

Against Docker Hub, we're already limited by the rate limiter, which will be a more effective limit than anything here anyhow.